### PR TITLE
CLI: Volume set option for gd2 cli

### PR DESCRIPTION
**Pull request by devyanikota**
Updates #205
Signed-off-by: Devyani Kota <devyanikota@gmail.com>

### CODE REVIEW

**Comment by devyanikota** [comment]
@aravindavk is there a way we can get the value of the Flag Option.
Am trying to extract the volume options entered by the user using os.Args[] (Line 229) because:
On extracting using Flags.Args(), it omits the first option key :/

Example:
`$./glustercli volume set testvol --option afr.eager-lock on foo bar`

Output of the created json:
`[on foo bar]`

The solution here can be to extract the value of the flag 'option' and append the rest of the array of options.

**Comment by aravindavk** [comment]
> Am trying to extract the volume options entered by the user using os.Args[] (Line 229) because:
On extracting using Flags.Args(), it omits the first option key :/

Why you need `--option`?

File: cli/cmd/volume.go [hunk 1]
```diff
@@ -202,13 +210,36 @@ var volumeGetCmd = &cobra.Command{
...
+
 var volumeSetCmd = &cobra.Command{
 	Use:   "set",
 	Short: helpVolumeSetCmd,
 	Run: func(cmd *cobra.Command, args []string) {
-		validateNArgs(cmd, 3, 3)
+		validateNArgs(cmd, 3, 0)
 		volname := cmd.Flags().Args()[0]
-		fmt.Println("SET:", volname)
+		options := os.Args[5:]
```
Comment (aravindavk) [inline cli/cmd/volume.go]: Do not take Args directly from os.Args, This will have raw arguments which are not parsed. Use cmd.Flags().Args()

File: cli/cmd/volume.go [hunk 2]
```diff
@@ -202,13 +210,36 @@ var volumeGetCmd = &cobra.Command{
...
+func volumeOptionJsonHandler(cmd *cobra.Command, volname string, options []string) error {
+	vopt := make(map[string]string)
+	for op,val := range options {
+		if op % 2 == 0 {
+			vopt[val] = options[op+1]
+		}else{
+			continue
+		}
+	}
+	var op = map[string]interface{}{"options": vopt}
```
Comment (aravindavk) [inline cli/cmd/volume.go]: Avoid interface where not required. Try,

var op = map[string]map[string]string

File: cli/cmd/volume.go [hunk 3]
```diff
@@ -64,8 +69,11 @@ func init() {
 	// Volume Delete
 	volumeCmd.AddCommand(volumeDeleteCmd)
 
-	volumeCmd.AddCommand(volumeGetCmd)
+    //Volume Set
+	volumeSetCmd.Flags().StringVarP(&flagSetCmdData, "data", "d", "", "Volume Options")
```
Comment (aravindavk) [inline cli/cmd/volume.go]: Why this is required?

File: cli/cmd/common.go [hunk 4]
```diff
@@ -23,7 +23,9 @@ func failure(msg string, err int) {
 func validateNArgs(cmd *cobra.Command, min int, max int) {
 	nargs := len(cmd.Flags().Args())
 	if nargs < min || (max != 0 && nargs > max) {
-		cmd.Usage()
-		os.Exit(1)
+		if nargs-1 % 2 == 0 {
```
Comment (aravindavk) [inline cli/cmd/common.go]: why this change here? This is only required in VolumeSet, move this to that func itself

File: cli/cmd/volume.go [hunk 5]
```diff
@@ -202,13 +210,36 @@ var volumeGetCmd = &cobra.Command{
 	},
 }
 
+func volumeOptionJsonHandler(cmd *cobra.Command, volname string, options []string) error {
+	vopt := make(map[string]string)
+	for op,val := range options {
+		if op % 2 == 0 {
+			vopt[val] = options[op+1]
+		}else{
+			continue
```
Comment (aravindavk) [inline cli/cmd/volume.go]: else statement not adding any value here, can be removed

File: cli/cmd/volume.go (hunk 1 above)
Comment (devyanikota) [reply cli/cmd/volume.go]: I tried using cmd.Flags().Args()
but, like I explained earlier:
$./glustercli volume set testvol --data afr.eager-lock on foo bar
The volume options array being created is [on foo bar].
The first volume option entered is being ommitted.
Hence, thought of extracting the value of the flag: 'data' and then append it to the array.

File: cli/cmd/volume.go (hunk 2 above)
Comment (devyanikota) [reply cli/cmd/volume.go]: Noted.

File: cli/cmd/volume.go (hunk 3 above)
Comment (devyanikota) [reply cli/cmd/volume.go]: um, I had named the flag as 'option' earlier, rectified and renamed : '--data'

File: cli/cmd/volume.go (hunk 5 above)
Comment (devyanikota) [reply cli/cmd/volume.go]: Right. ok. Noted.

File: cli/cmd/common.go (hunk 4 above)
Comment (devyanikota) [reply cli/cmd/common.go]: My bad here. Sorry. Rectified.

File: cli/cmd/volume.go (hunk 3 above)
Comment (aravindavk) [reply cli/cmd/volume.go]: Both of them are not required. Do not use any flags for volume set. Just accept all as arguments.

    $./glustercli volume set testvol afr.eager-lock on foo bar

File: cli/cmd/volume.go (hunk 1 above)
Comment (aravindavk) [reply cli/cmd/volume.go]: That is because of `--option` or `--data` flag. Both of them not required. Just use 

    $./glustercli volume set testvol afr.eager-lock on foo bar
